File: go/src/internal/eventstore/schema.go
package eventstore

import (
	"github.com/apache/arrow/go/v17/arrow"
)

// Schema is the canonical event schema; keep it in sync with the Python side.
// from_start_ms is ms elapsed since the meeting start; session_started is 0.
// The absolute start/end instants live in session_started/session_ended
// metadata under "timestamp_ms" (Unix ms).
var Schema = arrow.NewSchema([]arrow.Field{
	{Name: "meeting_id", Type: arrow.BinaryTypes.String, Nullable: false},
	{Name: "from_start_ms", Type: arrow.PrimitiveTypes.Int64, Nullable: false},
	{Name: "event_type", Type: arrow.BinaryTypes.String, Nullable: false},
	{Name: "display_name", Type: arrow.BinaryTypes.String, Nullable: true},
	{Name: "challenge_id", Type: arrow.BinaryTypes.String, Nullable: true},
	{Name: "question_id", Type: arrow.BinaryTypes.String, Nullable: true},
	{Name: "metadata", Type: arrow.BinaryTypes.String, Nullable: true},
}, nil)

File: py/src/ptrack_analytics/schema.py
"""Polars schemas for the canonical Parquet event log and the JSONL question
bank. Keep in sync with the Go side.

from_start_ms: ms elapsed since the meeting start (session_started is 0). The
absolute start/end instants live in session_started/session_ended metadata
under "timestamp_ms" (Unix ms).
"""

from __future__ import annotations

import polars as pl

EVENT_SCHEMA: dict[str, pl.DataType | type[pl.DataType]] = {
    "meeting_id": pl.String,
    "from_start_ms": pl.Int64,
    "event_type": pl.String,
    "display_name": pl.String,
    "challenge_id": pl.String,
    "question_id": pl.String,
    "metadata": pl.String,
}

QUESTIONS_SCHEMA: dict[str, pl.DataType | type[pl.DataType]] = {
    "question_id": pl.String,
    "auto_submitted": pl.Boolean,
    "question_type": pl.String,
    "prompt": pl.String,
    "choices": pl.List(pl.String),
    "correct_answer": pl.String,
    "match_mode": pl.String,
    "tolerance": pl.Float64,
}
